<?xml version="1.0" encoding="UTF-8"?>

<launch>

  <!-- _______________________THIS IS THE TELEOP PORTION___________________ -->
  <arg name="joy_config" default="ps3" />
  <arg name="joy_dev" default="/dev/input/js0" />

  <arg name="config_filepath" default="$(find jackal_teleop)/config/ps3.config.yaml" /> 

  <group ns="bluetooth_teleop">
    <node pkg="joy" type="joy_node" name="joy_node">
      <param name="dev" value="$(arg joy_dev)" />
      <param name="deadzone" value="0.3" />
      <param name="autorepeat_rate" value="20" />
    </node>

    <node pkg="jackal_teleop" name="teleop_twist_joy" type="teleop_node">
      <rosparam command="load" file="$(arg config_filepath)" />
    </node>
  </group>
  <!-- _____________________________________________________________________ -->

  <!-- _____________________ THIS IS THE CONTROL PORTION____________________ -->
  <arg name="enable_ekf" default="$(optenv ENABLE_EKF true)"/>

  <rosparam command="load" file="$(find jackal_teleop)/config/control.yaml" />

  <node name="controller_spawner" pkg="controller_manager" type="spawner"
        args="jackal_joint_publisher jackal_velocity_controller" />

  <group if="$(arg enable_ekf)" >
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_localization">
      <rosparam command="load" file="$(find jackal_control)/config/robot_localization.yaml" />
    </node>

    <arg name="gx5_enable" value="$(optenv JACKAL_GX5_IMU 0)"/>
    <arg name="microstrain_enable" value="$(optenv JACKAL_IMU_MICROSTRAIN 0)"/>
    <group if="$(eval gx5_enable | microstrain_enable)">
      <!-- Optionally load the configuration for the secondary GX5-family IMU -->
      <rosparam>
        ekf_localization:
          imu1: gx5/imu/data
          imu1_config: [false, false, false,
                      true, true, true,
                      false, false, false,
                      true, true, true,
                      false, false, false]
          imu1_differential: false
      </rosparam>
      <param name="ekf_localization/imu1" value="$(optenv JACKAL_IMU_MICROSTRAIN_NAME gx5)/imu/data"/>
    </group>
  </group>

  <node pkg="twist_mux" type="twist_mux" name="twist_mux">
    <rosparam command="load" file="$(find jackal_control)/config/twist_mux.yaml" />
    <remap from="cmd_vel_out" to="jackal_velocity_controller/cmd_vel"/>
  </node>

  <group if="$(optenv JACKAL_CONTROL_EXTRAS 0)" >
    <rosparam command="load" file="$(env JACKAL_CONTROL_EXTRAS_PATH)" />
  </group>
  <!-- ________________________________________________________________________ -->

</launch>
